ion-schema/src/main/kotlin/com/amazon/ionschema/model/Constraint.kt

```diff
@@ -0,0 +1,235 @@
+     * [ISL 2.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#any_of).
+     */
+    data class AnyOf(val types: TypeArgumentList) : Constraint
+
+    /**
+     * Represents the `byte_length` constraint.
+     * See relevant section in [ISL 1.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#byte_length) and
+     * [ISL 2.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#byte_length).
+     */
+    data class ByteLength(val range: DiscreteIntRange) : Constraint {
+        init {
+            range.start?.let { require(it >= 0) }
+        }
+    }
+
+    /**
+     * Represents the `codepoint_length` constraint.
+     * See relevant section in [ISL 1.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#codepoint_length) and
+     * [ISL 2.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#codepoint_length).
+     */
+    data class CodepointLength(val range: DiscreteIntRange) : Constraint {
+        init {
+            range.start?.let { require(it >= 0) }
+        }
+    }
+
+    /**
+     * Represents the `container_length` constraint.
+     * See relevant section in [ISL 1.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#container_length) and
+     * [ISL 2.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#container_length).
+     */
+    data class ContainerLength(val range: DiscreteIntRange) : Constraint {
+        init {
+            range.start?.let { require(it >= 0) }
+        }
+    }
+
+    /**
+     * Represents the `contains` constraint.
+     * See relevant section in [ISL 1.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#contains) and
+     * [ISL 2.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#contains).
+     */
+    data class Contains(val values: List<IonValue>) : Constraint
+
+    /**
+     * Represents the `element` constraint.
+     * See relevant section in [ISL 1.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#element) and
+     * [ISL 2.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#element).
+     *
+     * The [distinct] modifier is only supported for Ion Schema 2.0 and higher.
+     */
+    data class Element(val type: TypeArgument, val distinct: Boolean = false) : Constraint
+
+    /**
+     * Represents the `exponent` constraint, introduced in Ion Schema 2.0.
+     * See relevant section in [ISL 2.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#exponent).
+     */
+    data class Exponent(val range: DiscreteIntRange) : Constraint
+
+    /**
+     * Represents the `field_names` constraint, introduced in Ion Schema 2.0.
+     * See relevant section in [ISL 2.0 spec](https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#field_names).
+     */
+    data class FieldNames(val type: TypeArgument, val distinct: Boolean = false) : Constraint
```

Comment:
(More information) Since `Constraint` has data classes for both ISL1.0 and ISL 2.0 constraints construction, how do we ensure that ISL2.0 constraints are not used to construct an ISL 1.0 type/schema?

Reply:
It seems like you have a plan to provide builders that ensure this so I resolving this.